Enable update to unlimited swap

hqhq opened the pull request:
We actually have this feature in help doc:
$ docker update --help | grep swap
      --memory-swap bytes          Swap limit equal to memory plus swap: '-1' to enable unlimited swap

And almost have this support, just needs a small
fix to enable it.

Signed-off-by: Qiang Huang <h.huangqiang@huawei.com>

<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/docker/docker/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**
Fix `docker update --memory-swap -1 <cnt1>`

**- How I did it**
Make negative valid value in toContainerdResources.

**- How to verify it**
```
Terminal 1:
$ docker run -ti -m 300M --memory-swap 500M ubuntu bash

Terminal 2:
$ docker ps
got the container id
$ docker update --memory-swap -1 <id>

Terminal 1:
# cat /sys/fs/cgroup/memory/memory.memsw.limit_in_bytes
see if it's 18446744073709551615
```

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->


**- A picture of a cute animal (not mandatory but encouraged)**

thaJeztah commented:
ping @crosbymichael @tiborvass PTAL

@hqhq I'm being extra careful here, as we recently had issues due to the recent issues around `-1` not working correctly 😅, and a kernel bug in RHEL 7.2/CentOS 7.2 kernels

thaJeztah commented on daemon/update_linux.go:
```diff
@@ -23,7 +23,7 @@ func toContainerdResources(resources container.Resources) libcontainerd.Resource
 	r.CpusetMems = resources.CpusetMems
 	r.MemoryLimit = uint64(resources.Memory)
-	if resources.MemorySwap > 0 {
+	if resources.MemorySwap != 0 {
```
Should we make this explicitly

```go
if resources.MemorySwap == -1 || resources.MemorySwap > 0 {
```

hqhq replied:
I don't know if that's necessary, other invalid values are already covered in `verifyContainerSettings`, so there won't be other values here.

thaJeztah commented on daemon/update_linux.go:
```diff
@@ -23,7 +23,7 @@ func toContainerdResources(resources container.Resources) libcontainerd.Resource
 	r.MemoryLimit = uint64(resources.Memory)
-	if resources.MemorySwap > 0 {
+	if resources.MemorySwap != 0 {
 		r.MemorySwap = uint64(resources.MemorySwap)
```
Wondering what's the right thing to do here for `-1`. IIRC, the kernel itself accepts `-1` as a special value (even though it otherwise takes an `uint`.

hqhq replied:
Yes, cgroup API takes this input as string, so we'll convert the value to string in runc. TBH we're using some hacky here as:
```
uint64(-1) = 18446744073709551615
int64(18446744073709551615) = -1
```
After some back and forth in runtime-spec, we finally decide to use `int64` for all kinds of memory limits (I'm not quite convinced though), so maybe we can change all memory limit types to `int64` in Moby someday.

hqhq commented:
ping @thaJeztah @crosbymichael @tiborvass

justincormack commented:
Why are you making changes on Solaris and Unix support for this? They do not support any of these functions anyway...

hqhq commented:
@justincormack I was just adapting the new semantics there, all these functions were already there. And the semantic changes appliy to all of linux, solaris and unix if they have `memory` and `memorySwap` in resources even though some of them don't care about the difference between `0` and `-1`.

The old semantic is: if memory swap is not supported, set it to `-1` which means don't touch it. (Which is wrong)
The new semantic is: if memory swap is not supported, set to `0` which means don't touch it. If memory swap is set to `-1`, means it'll be set to unlimited.

hqhq commented:
ping @justincormack @vieux

hqhq committed:
Enable update to unlimited swap

We actually have this feature in help doc:
$ docker update --help | grep swap
--memory-swap bytes Swap limit equal to memory plus swap: '-1' to enable unlimited swap

And almost have this support, just needs a small
fix to enable it.

Signed-off-by: Qiang Huang <h.huangqiang@huawei.com>

hqhq committed:
Make sure system support it before adapt it

I suggested we should adapt container setting before
varifing, but it was objected. If we do it after, we
should still check if system support them.

Signed-off-by: Qiang Huang <h.huangqiang@huawei.com>

hqhq committed:
Fix create container with swap -1 on no-swap machine

This bug exist long ago, it's not exposed because we
didn't pass -1 of swap to containerd.

Signed-off-by: Qiang Huang <h.huangqiang@huawei.com>

hqhq commented:
ping @thaJeztah @justincormack @vdemeester @dnephin

vieux commented:
@tiborvass wdyt ? this would set a new trend in the CLI

vdemeester commented:
Seems like there has not been too much activity on this one for a long time.. :sweat: 
@hqhq I'll close for the time being. If you want to work back on it, please ping me, I'll re-open :wink:
